samplePDF/samplePDFFDBase.cpp

```diff
@@ -51,6 +41,168 @@ samplePDFFDBase::~samplePDFFDBase()
+    MACH3LOG_ERROR("Number of dimensions is {}", nDimensions);
+    MACH3LOG_ERROR("Check that an XVarStr has been given in the sample config");
+    throw MaCh3Exception(__FILE__, __LINE__);
+  } else{
+    MACH3LOG_INFO("Found {} dimensions for sample binning", nDimensions);
+  }
+  
+  //Sanity check that some binning has been specified
+  if(SampleXBins.size() == 0 && SampleYBins.size() == 0){
+    MACH3LOG_ERROR("No binning specified for either X or Y of sample binning, please add some binning to the sample config {}", SampleManager->GetFileName());
+    throw MaCh3Exception(__FILE__, __LINE__);
+  }
+  
+  //FD file info
+  if (!CheckNodeExists(SampleManager->raw(), "InputFiles", "mtupleprefix")){
+    MACH3LOG_ERROR("InputFiles:mtupleprefix not given in {}, please add this", SampleManager->GetFileName());
+  }
+  mtupleprefix = SampleManager->raw()["InputFiles"]["mtupleprefix"].as<std::string>();
+  mtuplesuffix = SampleManager->raw()["InputFiles"]["mtuplesuffix"].as<std::string>();
+  splineprefix = SampleManager->raw()["InputFiles"]["splineprefix"].as<std::string>();
+  splinesuffix = SampleManager->raw()["InputFiles"]["splinesuffix"].as<std::string>();
+  
+  for (auto const &osc_channel : SampleManager->raw()["SubSamples"]) {
+    mtuple_files.push_back(mtupleprefix+osc_channel["mtuplefile"].as<std::string>()+mtuplesuffix);
+    spline_files.push_back(splineprefix+osc_channel["splinefile"].as<std::string>()+splinesuffix);
+    sample_vecno.push_back(osc_channel["samplevecno"].as<int>());
+    sample_nutype.push_back(PDGToProbs(static_cast<NuPDG>(osc_channel["nutype"].as<int>())));
+    sample_oscnutype.push_back(PDGToProbs(static_cast<NuPDG>(osc_channel["oscnutype"].as<int>())));
+    sample_signal.push_back(osc_channel["signal"].as<bool>());
+  }
+
+  //Now get selection cuts
+  double low_bound = 0;
+  double up_bound = 0;
+  double KinematicParamter = 0;
+  std::vector<double> SelectionVec;
+  //Now grab the selection cuts from the manager
+  for ( auto const &SelectionCuts : SampleManager->raw()["SelectionCuts"]) {
+    SelectionStr.push_back(SelectionCuts["KinematicStr"].as<std::string>());
+    SelectionBounds.push_back(SelectionCuts["Bounds"].as<std::vector<double>>());
+    low_bound = SelectionBounds.back().at(0);
+    up_bound = SelectionBounds.back().at(1);
+    KinematicParamter = static_cast<double>(ReturnKinematicParameterFromString(SelectionCuts["KinematicStr"].as<std::string>()));
+    MACH3LOG_INFO("Adding cut on {} with bounds {} to {}", SelectionCuts["KinematicStr"].as<std::string>(), SelectionBounds.back().at(0), SelectionBounds.back().at(1));
+    SelectionVec = {KinematicParamter, low_bound, up_bound};
+    StoredSelection.push_back(SelectionVec);
+  }
+  NSelections = SelectionStr.size();
+
+  return;
+}
+
+void samplePDFFDBase::Initialise() {
+
+  //First grab all the information from your sample config via your manager
+  ReadSampleConfig();
+
+  //Now initialise all the variables you will need
+  Init();
```

Comment:
a function named `Init` in a function named `Initialise`, that looks like it should be part of the constructor, is confusing. It appears that this is the virtual call to tell subclasses to allocate memory that we are about to use. This tightly ordered and dependent initialisation sequence looks like it is neccessary, but it can be simplified.

Subclasses should perform their `Init` in their constructors, and then this `Initialise` function should be named `InitialiseBase` or something clear.

If the ordering could be uncoupled so that the subclass can initialise last, but still all be done in the subclass constructor, that would be nice, but I suspect it can't, so the flow here is almost the clearest it can be, but having functions named `Init` or similar that are just providing RAII/construction logic is confusing.

If you need multiple constructor signatures that differ but have common procedures, you can use delegating constructors: https://www.ibm.com/docs/en/zos/2.4.0?topic=only-delegating-constructors-c11

Reply:
See above comment reply. I agree that this needs simplifying!